### PR TITLE
Update `cupy.array_api`

### PR DESCRIPTION
**Pull request by asi1024**
This PR applies the patch from:
- numpy/numpy#20527
- numpy/numpy#20533
- numpy/numpy#20719
- numpy/numpy#20720
- numpy/numpy#20762
- numpy/numpy#20788
- numpy/numpy#20791
- numpy/numpy#20883

### CODE REVIEW

**Comment by leofang** [comment]
Thank you @asi1024 for filling in for me 😓

@asmeurer @rgommers this is something I forgot to raise last week: Can we set up a notification mechanism so that downstream projects can track any changes in `numpy.array_api` more easily, say a bot notifying a group of people? Currently we have had to check regularly the upstream change and, as you can see, it's easy to drop the ball (sorry!) ...

**Comment by kmaehashi** [comment]
`from cupy.array_api import _array_object` requires Python 3.8 and raises an warning.
How about implementing `__cupy_get_ndarray__` to Array? @asi1024

File: cupy/array_api/tests/test_data_type_functions.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import pytest
```
Comment (kmaehashi) [inline cupy/array_api/tests/test_data_type_functions.py:1]: Test code needs to live outside of `cupy/array_api`.

File: cupy/array_api/tests/test_set_functions.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import pytest
+from hypothesis import given
```
Comment (kmaehashi) [inline cupy/array_api/tests/test_set_functions.py:2]: As NumPy started to use hypothesis maybe we should consider using it too.
https://hypothesis.readthedocs.io/en/latest/index.html

**Comment by asmeurer** [comment]
Might be a question to ask on the numpy list if they can set up such a bot. I'm not a numpy core contributor so I don't have control of such things.

**Comment by rgommers** [comment]
> @asmeurer @rgommers this is something I forgot to raise last week: Can we set up a notification mechanism so that downstream projects can track any changes in `numpy.array_api` more easily, say a bot notifying a group of people? Currently we have had to check regularly the upstream change and, as you can see, it's easy to drop the ball (sorry!) ...

This is a problem we have in multiple places, for example in SciPy with people wanting to follow one specific module. We started using CODEOWNERS, but that's not a real fix, more like a band-aid. I'd like a bot (the PyTorch one is quite nice for example), except I don't really have bandwidth to maintain it (and it's one of those "need full permissions on the repo to fix problems"). Maybe we can find a volunteer though, I'll bring it up. In the meantime, would adding you to CODEOWNERS so you get pinged on all PRs work @leofang?

**Comment by leofang** [comment]
Sounds good. Thanks, Ralf. If I get notified but don't have bandwidth to address it immediately, I will relay it to the team.

**Comment by leofang** [comment]
(And yes, the PyTorch bot is quite nice, but it doesn't seem it's open sourced...?)

**Comment by rgommers** [comment]
> (And yes, the PyTorch bot is quite nice, but it doesn't seem it's open sourced...?)

It was https://github.com/pytorch/probot which has a permissive license. Just moved to https://github.com/pytorch/test-infra/tree/main/torchci which is missing a license - but that may be an oversight.

**Review by kmaehashi — DISMISSED** [review]

File: cupy/array_api/tests/test_set_functions.py (hunk 2 above)
Comment (honno) [reply cupy/array_api/tests/test_set_functions.py]: Feel free to ping me on any related issues/PRs if you need help with adopting Hypothesis.

**Comment by kmaehashi** [comment]
Test seems failing.
> FAILED cupy_tests/array_api_tests/test_set_functions.py::test_inverse_indices_shape[unique_all]

**Comment by asi1024** [comment]
@kmaehashi Now all tests have passed!

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Pull request merged**